### PR TITLE
Issue #2021 - Finalizer on an operator custom resource prevents the c…

### PR DESCRIPTION
**Pull request by MaxMcAdam**
…r and crd from being deleted

Signed-off-by: MaxMcAdam <max@fredcom.com>

### CODE REVIEW

**Comment by MaxMcAdam** [comment]
The deployment object handles the execution of finalizers. If the operator custom resource has a finalizer and the deployment object is uninstalled before the finalizer can finish executing, the finalizer will not be removed from the CR and the CR becomes unkillable until you manually patch the CR to remove the finalizer. This also blocks the CRD from being removed successfully.

File: kube_operator/api_objects.go [hunk 1]
```diff
@@ -456,6 +441,27 @@ func (cr CustomResourceV1Beta1) Uninstall(c KubeClient, namespace string) {
...
 	if err != nil {
 		glog.Warningf(kwlog(fmt.Sprintf("unable to delete operator custom resource %s. Error: %v", newCrName, err)))
+	} else {
+		status, err := cr.Status(c, namespace)
+		timeout := 20
+		for (err == nil || status != nil) && timeout > 0 {
+			glog.Infof(kwlog(fmt.Sprintf("Custom Resource %s is not yet down. Pausing for 10 before checking again. Custom resource status is: %v Error is %v", newCrName, status, err)))
+			time.Sleep(10 * time.Second)
+			timeout = timeout - 1
+		}
```
Comment (dabooz) [inline kube_operator/api_objects.go:451]: Put an error message in the log if the timeout occurs, and in the same code below too. You could make a reusable function for the whole status check too.

File: kube_operator/client.go [hunk 2]
```diff
@@ -333,7 +337,22 @@ func getK8sObjectFromYaml(yamlFiles []YamlFile, sch *runtime.Scheme) ([]APIObjec
 	_ = v1scheme.AddToScheme(sch)
 	_ = scheme.AddToScheme(sch)
 
-	for _, fileStr := range yamlFiles {
+	// multiple yaml files can be in one file separated by '---'
+	// these are split here and rejoined with the single files
```
Comment (dabooz) [inline kube_operator/client.go:341]: i really dislike yaml...the person who thought this was a good idea should have their coding license revoked.

**Review by dabooz — CHANGES_REQUESTED** [review]

File: kube_operator/api_objects.go (hunk 1 above)
Comment (MaxMcAdam) [reply kube_operator/api_objects.go:448]: Fixed in both

File: kube_operator/client.go (hunk 2 above)
Comment (MaxMcAdam) [reply kube_operator/client.go:341]: can't trust anything that puts so much syntactic significance on whitespace

**Review by dabooz — APPROVED** [review]

**Pull request merged**